In [2]:
import os 
import gym
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import DummyVecEnv
from stable_baselines3.common.evaluation import evaluate_policy

Matplotlib created a temporary cache directory at /var/folders/4s/rj_sy56d06508gj9z8g3lwhw0000gn/T/matplotlib-lq3xwtkb because the default path (/Users/yousuf/.matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.


In [3]:
env_name = 'MountainCar-v0'
env = gym.make(env_name, render_mode='human')

In [4]:
log_path = os.path.join('500k_PPO_Training_MC', 'Logs')

In [5]:
env = DummyVecEnv([lambda: env])
model = PPO('MlpPolicy', env, verbose=1, tensorboard_log=log_path)

/Users/yousuf/anaconda3/lib/python3.11/site-packages/stable_baselines3/common/vec_env/patch_gym.py:49: UserWarning: You provided an OpenAI Gym environment. We strongly recommend transitioning to Gymnasium environments. Stable-Baselines3 is automatically wrapping your environments in a compatibility layer, which could potentially cause issues.
  warnings.warn(


Using cpu device


In [ ]:
## Execute this 
## I have executed it in another ipynb file so I will just load that model here
model.learn(total_timesteps=100000)

### saving the model

In [6]:
test_env = gym.make(env_name, render_mode='human')

In [7]:
PPO_Path = os.path.join('Training_MC', 'Saved Models', '500k_Mountain_Car') 

In [ ]:
# execute this after trainning to save your model
model.save(PPO_Path)

In [8]:
model = PPO.load(PPO_Path, env=test_env)

Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


/Users/yousuf/anaconda3/lib/python3.11/site-packages/stable_baselines3/common/vec_env/patch_gym.py:49: UserWarning: You provided an OpenAI Gym environment. We strongly recommend transitioning to Gymnasium environments. Stable-Baselines3 is automatically wrapping your environments in a compatibility layer, which could potentially cause issues.
  warnings.warn(


### Testing the model

In [9]:
# Testing for 5 epochs
episodes = 5
for episode in range(1,episodes+1):
    obs, _ = test_env.reset()
    done=False
    score =0
    
    while not done:
        action, _ = model.predict(obs, deterministic=True)
        obs, reward, terminated, truncated, _ = test_env.step(action)
        done = terminated or truncated
        score += reward
        test_env.render()
        print('Episode:{} Score:{}'.format(episode, score))

test_env.close()

/Users/yousuf/anaconda3/lib/python3.11/site-packages/gym/utils/passive_env_checker.py:233: DeprecationWarning: `np.bool8` is a deprecated alias for `np.bool_`.  (Deprecated NumPy 1.24)
  if not isinstance(terminated, (bool, np.bool8)):


Episode:1 Score:-1.0
Episode:1 Score:-2.0
Episode:1 Score:-3.0
Episode:1 Score:-4.0
Episode:1 Score:-5.0
Episode:1 Score:-6.0
Episode:1 Score:-7.0
Episode:1 Score:-8.0
Episode:1 Score:-9.0
Episode:1 Score:-10.0
Episode:1 Score:-11.0
Episode:1 Score:-12.0
Episode:1 Score:-13.0
Episode:1 Score:-14.0
Episode:1 Score:-15.0
Episode:1 Score:-16.0
Episode:1 Score:-17.0
Episode:1 Score:-18.0
Episode:1 Score:-19.0
Episode:1 Score:-20.0
Episode:1 Score:-21.0
Episode:1 Score:-22.0
Episode:1 Score:-23.0
Episode:1 Score:-24.0
Episode:1 Score:-25.0
Episode:1 Score:-26.0
Episode:1 Score:-27.0
Episode:1 Score:-28.0
Episode:1 Score:-29.0
Episode:1 Score:-30.0
Episode:1 Score:-31.0
Episode:1 Score:-32.0
Episode:1 Score:-33.0
Episode:1 Score:-34.0
Episode:1 Score:-35.0
Episode:1 Score:-36.0
Episode:1 Score:-37.0
Episode:1 Score:-38.0
Episode:1 Score:-39.0
Episode:1 Score:-40.0
Episode:1 Score:-41.0
Episode:1 Score:-42.0
Episode:1 Score:-43.0
Episode:1 Score:-44.0
Episode:1 Score:-45.0
Episode:1 Score:-46

Episode:3 Score:-72.0
Episode:3 Score:-73.0
Episode:3 Score:-74.0
Episode:3 Score:-75.0
Episode:3 Score:-76.0
Episode:3 Score:-77.0
Episode:3 Score:-78.0
Episode:3 Score:-79.0
Episode:3 Score:-80.0
Episode:3 Score:-81.0
Episode:3 Score:-82.0
Episode:3 Score:-83.0
Episode:3 Score:-84.0
Episode:3 Score:-85.0
Episode:3 Score:-86.0
Episode:3 Score:-87.0
Episode:3 Score:-88.0
Episode:3 Score:-89.0
Episode:3 Score:-90.0
Episode:3 Score:-91.0
Episode:3 Score:-92.0
Episode:3 Score:-93.0
Episode:3 Score:-94.0
Episode:3 Score:-95.0
Episode:3 Score:-96.0
Episode:3 Score:-97.0
Episode:3 Score:-98.0
Episode:3 Score:-99.0
Episode:3 Score:-100.0
Episode:3 Score:-101.0
Episode:3 Score:-102.0
Episode:3 Score:-103.0
Episode:3 Score:-104.0
Episode:3 Score:-105.0
Episode:3 Score:-106.0
Episode:3 Score:-107.0
Episode:3 Score:-108.0
Episode:3 Score:-109.0
Episode:3 Score:-110.0
Episode:3 Score:-111.0
Episode:4 Score:-1.0
Episode:4 Score:-2.0
Episode:4 Score:-3.0
Episode:4 Score:-4.0
Episode:4 Score:-5.0
Epi

## The mountain Car Problem is considered solve at around -110 reward averaged over 100 runs

we are pretty much on point with it 